### PR TITLE
Add log probs feature to TT-Transformers on T3K

### LINKED ISSUE #32210: LLM Feature Request: logprobs
The `logprobs` parameter in the OpenAI chat completions API https://platform.openai.com/docs/api-reference/chat/create#chat_create-logprobs signals the model to return log probabilities for each output token. We need to enable the configuration of the logprobs parameter in our on-device sampling implementation, for each request in a batch. Note that the models should optionally return logprobs - with the caller making use of them if specified.

This issue serves as the parent issue for all tasks which are needed to expose this parameter.

| Model           | Log-Probs Enabled | Avg TSU | Time  |
|-----------------|-----------------|---------|-------|
| Qwen3 T3K       | No              |     22.14 t/s/u    |   45.16ms    |
| Qwen3 T3K       | Yes             |     16.29 t/s/u    |   61.40ms    |
| Llama3.1-8B     | No              |     48.83 t/s/u    |   20.48ms    |
| Llama3.1-8B     | Yes             |   29.75 t/s/u  |   33.61ms   |

### LINKED ISSUE #32262: Model: Add optional returning of log-softmaxed logits
Enable the model to optionally return log-softmaxed logits for each output token when requested. This will allow downstream use of logprobs for inference and analysis, and is needed to support the OpenAI-compatible logprobs API.

Requirements:
- Implement an option in the model to return log-softmaxed logits.
- Ensure this is configurable per request and only computed/returned when needed.
- Update documentation to reflect the optional output.
- Add tests to confirm correct output format and values.

Links:
Parent Issue: https://github.com/tenstorrent/tt-metal/issues/32210

### PR DESCRIPTION
**Pull request by djordje-tt**
### Ticket
#[32262](https://github.com/tenstorrent/tt-metal/issues/32262)

### Problem description
We are missing log-probs feature in tt-metal which is requirement for UF release eow.

### Feature description: 
For each sampled token by sampling op, it's required to return log-probability of it. It gives confidence of token that is sampled as extra information apart from produced token. 

### What's changed
- Implementation of LogProbsCalculator class for calculating log-probs
- Added unit tests with Qwen3 shapes
- Integration into simple_text_demo.py 
- To calculate it, LogProbsCalculator class is responsible for. All the implementation

Warning: Implementation of penalties here should be skipped reviewing since it's rebase to @sraizada-tt's branch: 

Review appreciated on: 
- `logProbsCalculator` class inside utils.py
- propagation in `generator.py`
- ignore function names in unit testing since it's still not polished. Mostly early opened to get early feedback on implementation and integration approach 
cc: @skhorasganiTT @sraizada-tt @rdraskicTT your feedback is much appreciated!

pipelines: 
- [BH Multi-card demo](https://github.com/tenstorrent/tt-metal/actions/runs/19853858206)
- [Galaxy Quick](https://github.com/tenstorrent/tt-metal/actions/runs/19854152113)
- [Galaxy Demo](https://github.com/tenstorrent/tt-metal/actions/runs/19854177690)
- [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/19853904911)
- [(T3K) demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/19853915344)

### CODE REVIEW

File: models/demos/llama3_70b_galaxy/demo/text_demo.py [hunk 1]
```diff
@@ -937,30 +937,30 @@ def test_demo_text(
...
+                is_cur_pos_sharded=is_cur_pos_sharded,
+                is_page_table_sharded=is_page_table_sharded,
+            )
+            read_events.append(read_event)
+            tt_out_toks.append(tt_out_tok)
+            if apc_test and iteration == 0:
+                tt_out_logits_saved_iter_0 = tt_out_logits_saved
+            # except Exception as e:
+            #     logger.error(f"Error during decoding: {str(e)}")
+            #     break
```
Comment (mtairum) [inline models/demos/llama3_70b_galaxy/demo/text_demo.py]: Why commenting this?

**Review by mtairum — APPROVED** [review]
Approving just to unblock for release this week.
Haven't done full review.

**Review by skhorasganiTT — COMMENTED** [review]
This PR is missing changes from https://github.com/tenstorrent/tt-metal/pull/32807, making it hard to review. Also it would be much preferred to merge that one first, not sure why it was closed. In the PR description could you also link the corresponding vLLM PR needed alongside these changes?

File: models/demos/llama3_70b_galaxy/demo/text_demo.py (hunk 1 above)
Comment (djordje-tt) [reply models/demos/llama3_70b_galaxy/demo/text_demo.py]: Returning back, was just testing something for llama.

[the remark this answers is not in the document]
Comment (djordje-tt) [reply models/demos/llama3_70b_galaxy/demo/text_demo.py]: Resolved!

**Comment by djordje-tt** [comment]
> This PR is missing changes from #32807, making it hard to review. Also it would be much preferred to merge that one first, not sure why it was closed. In the PR description could you also link the corresponding vLLM PR needed alongside these changes?

@skhorasganiTT, it's rebased now and cleaned up. @rdraskicTT will provide vllm PR for this.

File: models/common/utils.py [hunk 2]
```diff
@@ -45,3 +47,183 @@ def top_k_top_p_filtering(
...
+
+
+class LogProbsCalculator:
+    """
+    Class to calculate log-probs for a given logits tensor and indices tensor.
+
+    Args:
+        mesh_device: MeshDevice to use for all-gather operations
+        vocab_size: Vocabulary size
+    """
```
Comment (sraizada-tt) [inline models/common/utils.py:58]: can we put this class in a new file tt_lobprobs.py in the sampling folder and then import it in the sampling generator? 

and then call this in the _run_sampling method of the generator instead of in tt_sampling?

Comment (djordje-tt) [reply models/common/utils.py:58]: It is dependent on sampling so I'd rather keep it there. Since we can't get logprobs if we don't get token itself. Being that said, I think it's natural place should be inside sampling otherwise it does not exist when sampling doesn't.

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -568,17 +568,19 @@ def _decode_forward_no_trace_text(
...
                 tt_tokens[i],
                 tt_current_pos[i],
                 rot_mat_idxs=tt_rot_mat_idxs[i],
                 page_table=tt_page_table[i],
                 kv_cache=user_kv_cache,
                 sampling_on_device=sampling_on_device,
             )
-            tt_logits.append(tt_logits_i)
+            # tt_logits.append(tt_logits_i)
+            # tt_log_probs.append(tt_log_probs_i)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: please delete comments

File: models/tt_transformers/tt/generator.py [hunk 4]
```diff
@@ -1076,13 +1092,31 @@ def process_decode_output_host(self, tt_out, is_tokens=False):
...
+            elif isinstance(tt_out[i], ttnn.Tensor):
+                logits_i = self.model[i].process_output_decode(
+                    tt_out[i], max_batch_size_per_model, S=1, is_tokens=is_tokens
+                )
+                logits.append(logits_i)
+                # add dummy tensor for log_probs
+                log_probs.append(torch.ones(logits_i.shape))
+            else:
+                raise ValueError(f"Invalid type of tt_out: {type(tt_out[i])}")
+        return (torch.cat(logits, 0), torch.cat(log_probs, 0))
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:1149]: @rdraskicTT please ensure both the v0 and v1 model runners can handle both 1 or 2 outputs (1 for other models) in https://github.com/tenstorrent/vllm/pull/266. For v1 log prob outputs can be ignored for now.

**Review by skhorasganiTT — COMMENTED** [review]
@djordje-tt @rdraskicTT please also ensure vLLM nightly (https://github.com/tenstorrent/tt-metal/actions/workflows/vllm-nightly-tests.yaml) still passes for all models when run with the corresponding vLLM branch & that both PRs are ready to merge before merging either.

**Review by skhorasganiTT — APPROVED** [review]
generator lgtm assuming vllm nightly with corresponding vllm branch passes for all models (not just llama)

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_erfinv.h [hunk 5]
```diff
@@ -7,24 +7,23 @@
 #include "ckernel.h"
 #include "ckernel_defs.h"
 #include "ckernel_sfpu_log.h"
-
 #include "sfpi.h"
 
+using namespace sfpi;
```
Comment (fvranicTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_erfinv.h]: Could you tell us why we need this change? It looks like a revert.

Comment (djordje-tt) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_erfinv.h]: Oops, ignore all changes. Accidentally forgot to remove revert that was causing some accuracy issues.
This will be remove shortly!

Comment (fvranicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_erfinv.h]: Thanks!

**Review by gwangTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_sampling.py [hunk 6]
```diff
@@ -228,3 +230,67 @@ def test_sampling_subcores_callback(shape, k, p, seed, device, sub_core_grids):
...
+)
+@pytest.mark.parametrize(
+    "device_params",
+    [
+        ({"fabric_config": ttnn.FabricConfig.FABRIC_1D}),
+    ],
+    indirect=["device_params"],
+    ids=["fabric_linear"],
+)
+def test_log_probs_calculation(shape, mesh_device):
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/eltwise/test_sampling.py]: This is not testing an eltwise function, please find a more appropriate location to put this test, thanks 👍

File: models/common/sampling/generator.py [hunk 7]
```diff
@@ -106,7 +106,10 @@ def reset_output_state(self, tokens):
     # ---------------------------------------------------------------------
     def reset_sampling_params(self, sampling_params):
         self.tt_sampling.reset_params(
-            k=sampling_params.top_k, p=sampling_params.top_p, temp=sampling_params.temperature
+            k=sampling_params.top_k,
+            p=sampling_params.top_p,
+            temp=sampling_params.temperature,
+            enable_log_probs=sampling_params.enable_log_probs,
```
Comment (tchedaTT) [inline models/common/sampling/generator.py:114]: Are you sure this is enough when switching between logprobs / no logprobs, and enable_log_probs doesn't need to be a part of TraceKey? in LogProbsCalculator there is a regular `if` determining whether to do the calculation or return the saved output tensor, and it seems like it would be frozen the first time that function is traced.

File: models/common/sampling/generator.py
Comment (tchedaTT) [inline models/common/sampling/generator.py:1]: Please update the generator's return type throughout this file.

Comment (tchedaTT) [inline models/common/sampling/generator.py:1]: in reset_output_state(), tt_penalties.reset_output_tokens(tokens) doesn't handle the new output type

File: models/common/utils.py (hunk 2 above)
Comment (tchedaTT) [reply models/common/utils.py:58]: +1 please move from utils to modes/common/sampling/logprobs.py

File: models/common/utils.py [hunk 8]
```diff
@@ -45,3 +47,200 @@ def top_k_top_p_filtering(
...
+        indices_tensor: ttnn.Tensor,
+    ):
+        """
+        Calculate log-probs for a given logits tensor and indices tensor.
+        """
+        if not self.enable_log_probs:
+            return self.output_tensor
+
+        if self.mesh_device.get_num_devices() != 8:
+            return self.output_tensor
```
Comment (tchedaTT) [inline models/common/utils.py:244]: Are we sure we want to merge this before we support single device etc?

**Comment by tchedaTT** [comment]
By not calling the LogProbsCalculator in prefill, we are not supporting logprobs on the first token decoded. I am not sure if this is acceptable.

File: models/common/sampling/generator.py
Comment (djordje-tt) [reply models/common/sampling/generator.py:1]: That is outside the scope of this PR.

File: models/common/sampling/generator.py (hunk 7 above)
Comment (djordje-tt) [reply models/common/sampling/generator.py:114]: Added code for switching trace key. I planned on adding it in next patch due to many rebases during development but I ended up adding it now. You can just check latest commit if you want for those changes only!

File: models/common/utils.py (hunk 2 above)
Comment (djordje-tt) [reply models/common/utils.py:58]: Will move it in the next PR when I add changes for Llama! Just want to merge this asap if there are no breaking changes!

File: models/common/sampling/generator.py
Comment (djordje-tt) [reply models/common/sampling/generator.py:1]: I will add that in the next PR when adding log-probs for Llama3.3-70b. Tight in time for merging so if there are no major blockers, I'd like to merge it asap. T3K demo, vllm nightly and single card demos are all passing!

Comment (tchedaTT) [reply models/common/sampling/generator.py:1]: Are you sure? This PR changes the output type from tensor to a tuple. It unpacks the tuple to tensor in many places so as not to break pre-existing code - why not here? This breaks a previously working feature.

Comment (tchedaTT) [reply models/common/sampling/generator.py:1]: > when adding log-probs for Llama3.3-70b

If this PR is meant to only add limited support for logprobs, please add a comment detailing the limitations

File: models/tt_transformers/tt/model.py [hunk 9]
```diff
@@ -464,7 +473,7 @@ def ttnn_decode_forward(
             # Send output logits to DRAM so L1 is not reserved for ttnn tracing and can be used by subsequent operations
             tt_logits = ttnn.to_memory_config(tt_logits, ttnn.DRAM_MEMORY_CONFIG)
 
-        return tt_logits
+        return tt_logits, None  # TODO: Add log_probs_return_value
```
Comment (tchedaTT) [inline models/tt_transformers/tt/model.py]: As this gets executed only in the case of host sampling, I think we shouldn't return logprobs - the TODO shouldn't be here.

File: models/common/utils.py [hunk 10]
```diff
@@ -45,3 +47,200 @@ def top_k_top_p_filtering(
...
+    def _prepare_relevant_logits(self, logits_tensor: ttnn.Tensor, global_idx_tensor: ttnn.Tensor):
+        """
+        Prepare global idx tensor with correct values on all devices.
+        """
+        size_per_device = logits_tensor.shape[-1]
+
+        # convert global_idx_tensor to ttnn.TILE_LAYOUT
+        global_idx_tilized_tensor = ttnn.to_layout(global_idx_tensor, ttnn.TILE_LAYOUT)
+
+        # TODO: Raise an issue on this since for UINT_32 ttnn.div produces incorrect output (all zeros)
```
Comment (tchedaTT) [inline models/common/utils.py:176]: Please raise the issue and remove the TODO before we forget

File: models/common/utils.py [hunk 11]
```diff
@@ -45,3 +47,200 @@ def top_k_top_p_filtering(
...
+            cluster_axis=None,
+            topology=ttnn.Topology.Linear,
+        )
+        self.global_exp_sum = ttnn.sum(gathered_sum_exp_tensors, dim=-1, keepdim=True)
+
+        # reshape global_max and global_exp_sum to support same output shape as sampling output -> (1, 1, 1, 32)
+        self.global_max = ttnn.reshape(self.global_max, (1, 1, 1, 32))
+        self.global_exp_sum = ttnn.reshape(self.global_exp_sum, (1, 1, 1, 32))
+
+    def _prepare_relevant_logits(self, logits_tensor: ttnn.Tensor, global_idx_tensor: ttnn.Tensor):
```
Comment (tchedaTT) [inline models/common/utils.py:167]: I can't quite trace the logic here - please add a comment explaining how we select the logits here (which bit is responsible for selecting the top32 logits, and why it always contains the actually selected token as required for vllm)

File: models/common/sampling/generator.py [hunk 12]
```diff
@@ -126,11 +132,18 @@ def _validate_trace_inputs(self, slot, logits: ttnn.Tensor, tt_out_tok: Optional
...
+                raise ValueError(
+                    "The provided output tensor does not match the tensor used during trace capture. "
+                    "Call `reset_trace()` before tracing with new tensors."
+                )
+        else:
+            if tt_out_tok is not None and tt_out_tok is not slot["output"]:
+                raise ValueError(
+                    "The provided output tensor does not match the tensor used during trace capture. "
+                    "Call `reset_trace()` before tracing with new tensors."
+                )
```
Comment (tchedaTT) [inline models/common/sampling/generator.py:146]: TTSampling returns a tuple even if logprobs are disabled - is this necessary?

File: models/common/utils.py
Comment (tchedaTT) [inline models/common/utils.py:1]: LogProbsCalculator isn't really compatible with tracing curently - it doesn't always return the same tensor, so with tracing enabled in SamplingGenerator we will return slot["output"] which won't contain the new, correct logprobs.

File: models/common/sampling/generator.py (hunk 7 above)
Comment (tchedaTT) [reply models/common/sampling/generator.py:114]: Tracing still will not work as expected - started a separate thread.

File: models/common/sampling/generator.py
Comment (tchedaTT) [reply models/common/sampling/generator.py:1]: My bad - I meant update_output_tokens, which is called here https://github.com/tenstorrent/tt-metal/blob/76e7f42ac80330f52934f8846635d9921ec15341/models/common/sampling/generator.py#L245

File: models/common/utils.py (hunk 11 above)
Comment (tchedaTT) [reply models/common/utils.py:167]: Answer: this version only returns the logprobs for the token that was actually sampled

File: models/common/utils.py
Comment (tchedaTT) [reply models/common/utils.py:1]: Seemingly because the input tensors to logprobscalculator are fixed, the output tensor is fixed as well, and the tracing works correctly. I am unable to confirm if this is guaranteed to work, but testing indicates it does.

File: models/common/utils.py (hunk 8 above)
Comment (djordje-tt) [reply models/common/utils.py:244]: Yes, this is P0 for UF customer that should be ready for eow so support for single device is definitely out of scope for this task now.

File: models/tt_transformers/tt/model.py (hunk 9 above)
Comment (djordje-tt) [reply models/tt_transformers/tt/model.py]: We are returning to fulfill fixed number of retvals for method. It's bad practice to return variable number of values.

File: models/common/sampling/generator.py
Comment (djordje-tt) [reply models/common/sampling/generator.py:1]: Updated PR name
